docarray/index/backends/milvus.py

```diff
@@ -0,0 +1,328 @@
+            )
+
+        for column_name, info in self._column_infos.items():
+            if info.n_dim is None and safe_issubclass(
+                info.docarray_type, AbstractTensor
+            ):
+                raise ValueError(
+                    f"Dimension information is missing for column '{column_name}' which is vector type"
+                )
+
+    def _build_index(self):
+        index = {
+            "index_type": self._db_config.index_type,
+            "metric_type": self._db_config.index_metric,
+            "params": self._db_config.index_params,
+        }
+
+        self._collection.create_index(self._db_config.index_name, index)
+        self._logger.info(
+            f"Index '{self._db_config.index_name}' has been successfully created"
+        )
+
+    def index(self, docs: Union[BaseDoc, Sequence[BaseDoc]], **kwargs):
+        """Index Documents into the index.
+
+        !!! note
+            Passing a sequence of Documents that is not a DocList
+            (such as a List of Docs) comes at a performance penalty.
+            This is because the Index needs to check compatibility between itself and
+            the data. With a DocList as input this is a single check; for other inputs
+            compatibility needs to be checked for every Document individually.
+
+        :param docs: Documents to index.
+        """
+
+        docs = self._validate_docs(docs)
+        entities = [[] for _ in range(len(self._column_infos.items()))]
+
+        for i in range(len(docs)):
+            for j, (column_name, info) in enumerate(self._column_infos.items()):
+                column_value = docs[i].__getattr__(column_name)
+                if isinstance(column_value, (tf.Tensor, torch.Tensor, np.ndarray)):
+                    column_value = self._convert_to_vector(column_value)
+                entities[j].append(column_value)
+
+        self._collection.insert(entities)
+        self._collection.flush()
+        self._logger.info(f"{len(docs)} documents has been indexed")
+
+    def _convert_to_vector(self, column_value: AbstractTensor) -> Sequence[float]:
+        """
+        Converts a column value to a float vector.
+
+        The database can only store float vectors, so this method is used to convert
+        TensorFlow or PyTorch tensors to a format compatible with the database.
+
+        :param column_value: The column value to be converted.
+        :return: The converted float vector.
+        """
+        if len(column_value.shape) != 1:
+            raise ValueError(
+                'Unsupported: Milvus backend only supports one-dimensional vectors'
+            )
+
+        if isinstance(column_value, np.ndarray):
+            return column_value.astype(float).tolist()
+        elif torch.is_tensor(column_value):
+            return column_value.float().numpy().tolist()
+        elif tf.is_tensor(column_value):
```

Comment:
This is not safe, because not every user has torch and/or tf installed. Therefore we can't directly use torch/tf. [Here](https://github.com/docarray/docarray/blob/3fc6ecb71bdc0095f2c405c17492debcc3d8412d/docarray/typing/tensor/video/video_tensor.py#L19) you can see how that is handled in other places.